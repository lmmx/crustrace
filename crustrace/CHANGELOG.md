# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.4](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.3...crustrace-v0.1.4) - 2025-06-09

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.3](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.2...crustrace-v0.1.3) - 2025-06-08

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.2](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.1...crustrace-v0.1.2) - 2025-06-07

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.1](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.0...crustrace-v0.1.1) - 2025-06-07

### <!-- 2 -->Bug Fixes

- amend README metadata

### <!-- 9 -->Other

- license

## [0.1.0](https://github.com/lmmx/crustrace/releases/tag/crustrace-v0.1.0) - 2025-05-27

### <!-- 1 -->Features

- support impl Struct methods
- Initial commit

### <!-- 5 -->Refactor

- rename 'trace all' to 'omni', it's cleaner
- merge crustrace-attributes crate into crustrace-core ([#3](https://github.com/lmmx/crustrace/pull/3))

### <!-- 8 -->Styling

- lint

### <!-- 9 -->Other

- prepare for publishing crustrace crate
- set up git hooks
