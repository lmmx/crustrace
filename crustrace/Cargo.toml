[package]
description = "Function and module-level procedural macro attributes to instrument functions with tracing"
documentation = "https://docs.rs/crustrace"
edition = "2024"
homepage = "https://github.com/lmmx/crustrace"
license = "MIT OR Apache-2.0"
name = "crustrace"
readme = "../README.md"
repository = "https://github.com/lmmx/crustrace"
version = "0.1.0"

[lib]
proc-macro = true

[dependencies]
crustrace-core = { workspace = true }
proc-macro2.workspace = true

[dev-dependencies]
cargo-husky = { workspace = true, features = ["user-hooks"] }
insta.workspace = true
rust-format.workspace = true
tracing.workspace = true
tracing-subscriber.workspace = true
