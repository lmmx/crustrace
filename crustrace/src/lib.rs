use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;

#[proc_macro_attribute]
pub fn trace_all(_args: TokenStream, input: TokenStream) -> TokenStream {
    let input2: TokenStream2 = input.into();
    let output = crustrace_core::trace_all_impl(input2);
    output.into()
}