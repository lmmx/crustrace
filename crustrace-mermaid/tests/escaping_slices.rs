// tests/escaping_slices.rs
#[cfg(test)]
mod escaping_slices_tests {
    use crustrace::instrument;
    use crustrace_mermaid::*;
    use tracing::subscriber::set_default;
    use tracing_subscriber::prelude::*;

    /// Function that takes references and slices to reproduce
    /// the `&[]` / `&["document"]` style output seen in Mermaid.
    #[instrument]
    fn slice_cases(name: &str, avro_schema: &[&str], record_stack: &[&str]) {
        let _ = (name, avro_schema, record_stack);
    }

    #[test]
    fn snapshot_slice_cases() {
        let layer = MermaidLayer::new().without_auto_flush();
        let subscriber = tracing_subscriber::registry().with(layer.clone());
        let _guard = set_default(subscriber);

        // Case 1: non-empty slice
        slice_cases("document", &["field1", "field2"], &["document"]);

        // Case 2: empty slice
        slice_cases("empty_case", &[], &[]);

        insta::assert_snapshot!(layer.render());
    }
}
