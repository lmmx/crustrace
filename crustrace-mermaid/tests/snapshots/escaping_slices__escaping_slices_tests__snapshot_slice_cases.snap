---
source: crustrace-mermaid/tests/escaping_slices.rs
expression: layer.render()
---
flowchart TD
subgraph Params1[" "]
  P2_0["name = document"]:::data
  P2_1["avro_schema = &#91;&quot;field1&quot;, &quot;field2&quot;]"]:::data
  P2_0 --- P2_1
  P2_2["record_stack = &#91;&quot;document&quot;]"]:::data
  P2_1 --- P2_2
end
F1["slice_cases()"]:::func
Params1 --> F1
subgraph Params2[" "]
  P3_0["name = empty_case"]:::data
  P3_1["avro_schema = &#91;]"]:::data
  P3_0 --- P3_1
  P3_2["record_stack = &#91;]"]:::data
  P3_1 --- P3_2
end
F2["slice_cases()"]:::func
Params2 --> F2

classDef func fill:#c6f6d5,stroke:#2f855a,stroke-width:2px,color:#22543d;
classDef data fill:#bee3f8,stroke:#2b6cb0,stroke-width:1.5px,color:#1a365d;
classDef params fill:none,stroke:#e53e3e,stroke-width:2px,color:#742a2a;
class Params1,Params2 params;
